fuel/server.py

```diff
@@ -0,0 +1,177 @@
+import logging
+from collections import deque
+from multiprocessing import Process
+
+import numpy
+import six
+import zmq
+
+from fuel.utils import buffer_
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level='INFO')
+
+
+def send_arrays(socket, arrays, flags=0, copy=True, track=False, stop=False):
+    """Send a NumPy array using the buffer interface and some metadata.
+
+    Parameters
+    ----------
+    socket : :class:`zmq.Socket`
+        The socket to send data over.
+    arrays : list
+        A list of :class:`numpy.ndarray` to transfer.
+    stop : bool, optional
+        Instead of sending a series of NumPy arrays, send a JSON object
+        with a single `stop` key. The :func:`recv_arrays` will raise
+        ``StopIteration`` when it receives this.
+
+    See Also
+    --------
+    :meth:`zmq.Socket.send` for other arguments
+
+    Notes
+    -----
+    The protocol is very simple: A single JSON object that contains the
+    shape and data type for each array is transferred first. Subsequently
+    the arrays are sent as bytestreams (through NumPy's support of the
+    buffering protocol).
+
+    """
+    if stop:
+        mds = {'stop': True}
+        return socket.send_json(mds, flags)
+    else:
+        mds = [{'dtype': str(array.dtype), 'shape': array.shape}
```

Comment:
I'd add two things:

* A C-ordered boolean flag, along with a check making sure this is the case that raises `NotImplementedError` for now (we don't necessarily need to support it, ever, but if someone is using some exotic library it could obviate some truly bizarre bugs). It's the one piece of metadata that the [NPY format](https://github.com/numpy/numpy/blob/master/doc/neps/npy-format.rst) stores that we're currently missing here, I think.
* Perhaps a format specifier that says that this is a NumPy array in the first place. This would future-proof it against a possible scenario where we want to send e.g. various types of `scipy.sparse` objects instead.